lib/charms/mysql/v0/mysql.py

```diff
@@ -311,6 +330,242 @@ class RouterUser:
     router_id: str
 
 
+class MySQLCharmBase(CharmBase):
+    """Base class to encapsulate charm related functionality.
+
+    Meant as a means to share common charm related code between the MySQL VM and
+    K8s charms.
+    """
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.app_secrets, self.unit_secrets = None, None
+
+    def _on_get_password(self, event: ActionEvent) -> None:
+        """Action used to retrieve the system user's password."""
+        username = event.params.get("username") or ROOT_USERNAME
+
+        if username not in REQUIRED_USERNAMES:
+            event.fail(
+                f"The action can be run only for users used by the charm: {', '.join(REQUIRED_USERNAMES)} not {username}"
+            )
+            return
+
+        if username == ROOT_USERNAME:
+            secret_key = ROOT_PASSWORD_KEY
+        elif username == SERVER_CONFIG_USERNAME:
+            secret_key = SERVER_CONFIG_PASSWORD_KEY
+        elif username == CLUSTER_ADMIN_USERNAME:
+            secret_key = CLUSTER_ADMIN_PASSWORD_KEY
+        elif username == MONITORING_USERNAME:
+            secret_key = MONITORING_PASSWORD_KEY
+        elif username == BACKUPS_USERNAME:
+            secret_key = BACKUPS_PASSWORD_KEY
+        else:
+            raise RuntimeError("Invalid username.")
+
+        event.set_results({"username": username, "password": self.get_secret("app", secret_key)})
```

Comment:
and this block can be:
```suggestion
        if secret_key := REQUIRED_USERS.get(username):
            event.set_results({"username": username, "password": self.get_secret("app", secret_key)})
        else:
            event.fail("Invalid username")
```